{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Bioinformatics with Jupyter Notebooks for WormBase:**\n",
    "## **Data Retrieval 2 - Getting data from WormMine**\n",
    "Welcome to the second jupyter notebook in the WormBase tutorial series. Over this series of tutorials, we will write code in Python that allows us to retrieve and perform simple analyses with data available on the WormBase sites.\n",
    "\n",
    "This tutorial will deal with the WormBase data from WormMine. (http://intermine.wormbase.org/tools/wormmine/begin.do)\n",
    "We will both explore the site, and the intermine python package, and extract data of interest. Let's get started!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We start by installing and loading the libraries that are required for this tutorial. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: intermine in /home/prajna/anaconda3/lib/python3.8/site-packages (1.11.0)\r\n"
     ]
    }
   ],
   "source": [
    "!pip install intermine\n",
    "import intermine\n",
    "from intermine import registry\n",
    "from intermine.webservice import Service\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "getInfo(mine) can fetch all the information about a particular mine i.e., its description, version, organisms associated etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Description: Intermine data mining platform for C. elegans and related nematodes\n",
      "URL: http://intermine.wormbase.org/tools/wormmine/\n",
      "API Version: 31\n",
      "Release Version: 280\n",
      "InterMine Version: 4.2.0\n",
      "Organisms: \n",
      "C. elegans\n",
      "Neighbours: \n",
      "Animals\n",
      "AGR\n"
     ]
    }
   ],
   "source": [
    "registry.getInfo(\"WormMine\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "getData(mine) can be used to extract the data sets corresponding to it.\n",
    "Note that this is not completely representative of the data available on WormMine!! Check the WormMine site for the complete data!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: C. elegans genomic annotations (GFF3 Gene)\n",
      "Name: Disease Ontology\n",
      "Name: FlyMine intergenic regions\n",
      "Name: GO\n",
      "Name: GO Annotation data set\n",
      "Name: Panther orthologue and paralogue predictions\n",
      "Name: WormBase B. malayi protein sequence (Fasta)\n",
      "Name: WormBase C. brenneri protein sequence (Fasta)\n",
      "Name: WormBase C. briggsae protein sequence (Fasta)\n",
      "Name: WormBase C. elegans CDS sequence (Fasta)\n",
      "Name: WormBase C. elegans chromosome sequence (Fasta)\n",
      "Name: WormBase C. elegans protein sequence (Fasta)\n",
      "Name: WormBase C. elegans transcript sequence (Fasta)\n",
      "Name: WormBase C. japonica protein sequence (Fasta)\n",
      "Name: WormBase C. remanei protein sequence (Fasta)\n",
      "Name: WormBase O. volvulus protein sequence (Fasta)\n",
      "Name: WormBase P. pacificus protein sequence (Fasta)\n",
      "Name: WormBase S. ratti protein sequence (Fasta)\n",
      "Name: WormBase T. muris protein sequence (Fasta)\n",
      "Name: WormBaseAcedbConverter\n",
      "Name: anatomy\n"
     ]
    }
   ],
   "source": [
    "registry.getData(\"WormMine\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The method \"new_query\" from Service class creates a query object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "service = Service(\"http://intermine.wormbase.org/tools/wormmine/service\")\n",
    "query=service.new_query()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Performing simple queries on the WormMine database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's query the WormMine database to extract the commonName, genus, name, shortName, species, and taxonID of all organisms whose data is available:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>commonName</th>\n",
       "      <th>genus</th>\n",
       "      <th>name</th>\n",
       "      <th>shortName</th>\n",
       "      <th>species</th>\n",
       "      <th>taxonId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Norway rat</td>\n",
       "      <td>Rattus</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>R. norvegicus</td>\n",
       "      <td>norvegicus</td>\n",
       "      <td>10116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>agent of lymphatic filariasis</td>\n",
       "      <td>Brugia</td>\n",
       "      <td>Brugia malayi</td>\n",
       "      <td>B. malayi</td>\n",
       "      <td>malayi</td>\n",
       "      <td>6279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>baker's yeast</td>\n",
       "      <td>Saccharomyces</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>S. cerevisiae</td>\n",
       "      <td>cerevisiae</td>\n",
       "      <td>4932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>barber pole worm</td>\n",
       "      <td>Haemonchus</td>\n",
       "      <td>Haemonchus contortus</td>\n",
       "      <td>H. contortus</td>\n",
       "      <td>contortus</td>\n",
       "      <td>6289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fruit fly</td>\n",
       "      <td>Drosophila</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>D. melanogaster</td>\n",
       "      <td>melanogaster</td>\n",
       "      <td>7227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>house mouse</td>\n",
       "      <td>Mus</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>M. musculus</td>\n",
       "      <td>musculus</td>\n",
       "      <td>10090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>human</td>\n",
       "      <td>Homo</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>H. sapiens</td>\n",
       "      <td>sapiens</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>pig roundworm</td>\n",
       "      <td>Ascaris</td>\n",
       "      <td>Ascaris suum</td>\n",
       "      <td>A. suum</td>\n",
       "      <td>suum</td>\n",
       "      <td>6253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>pine wood nematode</td>\n",
       "      <td>Bursaphelenchus</td>\n",
       "      <td>Bursaphelenchus xylophilus</td>\n",
       "      <td>B. xylophilus</td>\n",
       "      <td>xylophilus</td>\n",
       "      <td>6326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>roundworm</td>\n",
       "      <td>Caenorhabditis</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>C. elegans</td>\n",
       "      <td>elegans</td>\n",
       "      <td>6239</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      commonName            genus                        name  \\\n",
       "0                     Norway rat           Rattus           Rattus norvegicus   \n",
       "1  agent of lymphatic filariasis           Brugia               Brugia malayi   \n",
       "2                  baker's yeast    Saccharomyces    Saccharomyces cerevisiae   \n",
       "3               barber pole worm       Haemonchus        Haemonchus contortus   \n",
       "4                      fruit fly       Drosophila     Drosophila melanogaster   \n",
       "5                    house mouse              Mus                Mus musculus   \n",
       "6                          human             Homo                Homo sapiens   \n",
       "7                  pig roundworm          Ascaris                Ascaris suum   \n",
       "8             pine wood nematode  Bursaphelenchus  Bursaphelenchus xylophilus   \n",
       "9                      roundworm   Caenorhabditis      Caenorhabditis elegans   \n",
       "\n",
       "         shortName       species taxonId  \n",
       "0    R. norvegicus    norvegicus   10116  \n",
       "1        B. malayi        malayi    6279  \n",
       "2    S. cerevisiae    cerevisiae    4932  \n",
       "3     H. contortus     contortus    6289  \n",
       "4  D. melanogaster  melanogaster    7227  \n",
       "5      M. musculus      musculus   10090  \n",
       "6       H. sapiens       sapiens    9606  \n",
       "7          A. suum          suum    6253  \n",
       "8    B. xylophilus    xylophilus    6326  \n",
       "9       C. elegans       elegans    6239  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query(\"Organism\")\n",
    "query.select(\"commonName\", \"genus\", \"name\", \"shortName\", \"species\",\"taxonId\")\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "organisms_data = pd.DataFrame(columns = [\"commonName\", \"genus\", \"name\", \"shortName\", \"species\",\"taxonId\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'commonName':row[0], 'genus':row[1], 'name':row[2], 'shortName':row[3], 'species':row[4], \n",
    "            'taxonId':row[5]}\n",
    "    organisms_data = organisms_data.append(info, ignore_index = True)\n",
    "organisms_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's query the WormMine database to extract the automatedDescription, biotype, briefDescription, length, operon, primaryIdentifier, secondaryIdentifier, and symbol of all genes whose data is available:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>automatedDescription</th>\n",
       "      <th>biotype</th>\n",
       "      <th>briefDescription</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>primaryIdentifier</th>\n",
       "      <th>secondaryIdentifier</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>'Skinny hedgehog' (SKI1) encodes an enzyme tha...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>HGNC:18270</td>\n",
       "      <td>None</td>\n",
       "      <td>Hhat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(6-4)-photolyase (phr6-4) encodes an enzyme th...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>FB:FBgn0016054</td>\n",
       "      <td>None</td>\n",
       "      <td>phr6-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1,3-beta-glucanosyltransferase; has similarity...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SGD:S000005390</td>\n",
       "      <td>None</td>\n",
       "      <td>Gas5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1,3-beta-glucanosyltransferase; involved with ...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SGD:S000005492</td>\n",
       "      <td>None</td>\n",
       "      <td>Gas4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1,3-beta-glucanosyltransferase; involved with ...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SGD:S000004335</td>\n",
       "      <td>None</td>\n",
       "      <td>Gas2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1,4-Alpha-Glucan Branching Enzyme (AGBE) encod...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>FB:FBgn0053138</td>\n",
       "      <td>None</td>\n",
       "      <td>AGBE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1-acyl-sn-glycerol-3-phosphate acyltransferase...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SGD:S000002210</td>\n",
       "      <td>None</td>\n",
       "      <td>SLC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1-phosphatidylinositol-3-phosphate 5-kinase; v...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SGD:S000001915</td>\n",
       "      <td>None</td>\n",
       "      <td>FAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>14-3-3 protein, major isoform; controls proteo...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SGD:S000000979</td>\n",
       "      <td>None</td>\n",
       "      <td>BMH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>14-3-3 protein, minor isoform; controls proteo...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>SGD:S000002506</td>\n",
       "      <td>None</td>\n",
       "      <td>BMH2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                automatedDescription     biotype  \\\n",
       "0  'Skinny hedgehog' (SKI1) encodes an enzyme tha...  SO:0001217   \n",
       "1  (6-4)-photolyase (phr6-4) encodes an enzyme th...  SO:0001217   \n",
       "2  1,3-beta-glucanosyltransferase; has similarity...  SO:0001217   \n",
       "3  1,3-beta-glucanosyltransferase; involved with ...  SO:0001217   \n",
       "4  1,3-beta-glucanosyltransferase; involved with ...  SO:0001217   \n",
       "5  1,4-Alpha-Glucan Branching Enzyme (AGBE) encod...  SO:0001217   \n",
       "6  1-acyl-sn-glycerol-3-phosphate acyltransferase...  SO:0001217   \n",
       "7  1-phosphatidylinositol-3-phosphate 5-kinase; v...  SO:0001217   \n",
       "8  14-3-3 protein, major isoform; controls proteo...  SO:0001217   \n",
       "9  14-3-3 protein, minor isoform; controls proteo...  SO:0001217   \n",
       "\n",
       "  briefDescription length operon primaryIdentifier secondaryIdentifier  symbol  \n",
       "0             None   None   None        HGNC:18270                None    Hhat  \n",
       "1             None   None   None    FB:FBgn0016054                None  phr6-4  \n",
       "2             None   None   None    SGD:S000005390                None    Gas5  \n",
       "3             None   None   None    SGD:S000005492                None    Gas4  \n",
       "4             None   None   None    SGD:S000004335                None    Gas2  \n",
       "5             None   None   None    FB:FBgn0053138                None    AGBE  \n",
       "6             None   None   None    SGD:S000002210                None    SLC1  \n",
       "7             None   None   None    SGD:S000001915                None    FAB1  \n",
       "8             None   None   None    SGD:S000000979                None    BMH1  \n",
       "9             None   None   None    SGD:S000002506                None    BMH2  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query(\"Gene\")\n",
    "query.select(\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \"primaryIdentifier\", \n",
    "             \"secondaryIdentifier\", \"symbol\")\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "genes_data = pd.DataFrame(columns = [\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \n",
    "                                     \"primaryIdentifier\", \"secondaryIdentifier\", \"symbol\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'automatedDescription':row[0], 'biotype':row[1], 'briefDescription':row[2], 'length':row[3], \n",
    "            'operon':row[4], 'primaryIdentifier':row[5], 'secondaryIdentifier':row[6], 'symbol':row[7]}\n",
    "    genes_data = genes_data.append(info, ignore_index = True)\n",
    "genes_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's create a query object to query the description of all Gene Ontology (GO) terms from the WormMine database. We can then add extra columns (fields) to our query based on our needs - here, the identifier for all GO terms is added to the query. \n",
    "When there are multiple fields in a query, the default ordering of the output is based on the first field, but this can be changed to any other field."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GOTerm.description</th>\n",
       "      <th>GOTerm.identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>The distribution of mitochondria, including th...</td>\n",
       "      <td>GO:0000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>The maintenance of the structure and integrity...</td>\n",
       "      <td>GO:0000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>The production of new individuals that contain...</td>\n",
       "      <td>GO:0000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Enables the transfer of zinc ions (Zn2+) from ...</td>\n",
       "      <td>GO:0000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Enables the transfer of a solute or solutes fr...</td>\n",
       "      <td>GO:0000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Catalysis of the transfer of a mannose residue...</td>\n",
       "      <td>GO:0000009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Catalysis of the reaction: all-trans-hexapreny...</td>\n",
       "      <td>GO:0000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>The distribution of vacuoles into daughter cel...</td>\n",
       "      <td>GO:0000011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>The repair of single strand breaks in DNA. Rep...</td>\n",
       "      <td>GO:0000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Catalysis of the hydrolysis of ester linkages ...</td>\n",
       "      <td>GO:0000014</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  GOTerm.description GOTerm.identifier\n",
       "0  The distribution of mitochondria, including th...        GO:0000001\n",
       "1  The maintenance of the structure and integrity...        GO:0000002\n",
       "2  The production of new individuals that contain...        GO:0000003\n",
       "3  Enables the transfer of zinc ions (Zn2+) from ...        GO:0000006\n",
       "4  Enables the transfer of a solute or solutes fr...        GO:0000007\n",
       "5  Catalysis of the transfer of a mannose residue...        GO:0000009\n",
       "6  Catalysis of the reaction: all-trans-hexapreny...        GO:0000010\n",
       "7  The distribution of vacuoles into daughter cel...        GO:0000011\n",
       "8  The repair of single strand breaks in DNA. Rep...        GO:0000012\n",
       "9  Catalysis of the hydrolysis of ester linkages ...        GO:0000014"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query()\n",
    "query.select(\"GOTerm.description\")\n",
    "\n",
    "query.add_view(\"GOTerm.identifier\") #Add a column to the query\n",
    " \n",
    "query.add_sort_order(\"GOTerm.identifier\") #Changing the sorting order of the query by a specific column\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "GO_data = pd.DataFrame(columns = [\"GOTerm.description\", \"GOTerm.identifier\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'GOTerm.description':row[0], 'GOTerm.identifier':row[1]}\n",
    "    GO_data = GO_data.append(info, ignore_index = True)\n",
    "GO_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Performing queries on the WormMine database using constraints"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's query the WormMine database to extract all details of all organisms for which data is available. Then we can add a constraint to the query to restrict outputs to only those with a certain value in one of the fields, here Caenorhabditis in genus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>commonName</th>\n",
       "      <th>genus</th>\n",
       "      <th>name</th>\n",
       "      <th>shortName</th>\n",
       "      <th>species</th>\n",
       "      <th>taxonId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>roundworm</td>\n",
       "      <td>Caenorhabditis</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>C. elegans</td>\n",
       "      <td>elegans</td>\n",
       "      <td>6239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>Caenorhabditis</td>\n",
       "      <td>Caenorhabditis angaria</td>\n",
       "      <td>C. angaria</td>\n",
       "      <td>angaria</td>\n",
       "      <td>860376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>Caenorhabditis</td>\n",
       "      <td>Caenorhabditis brenneri</td>\n",
       "      <td>C. brenneri</td>\n",
       "      <td>brenneri</td>\n",
       "      <td>135651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>Caenorhabditis</td>\n",
       "      <td>Caenorhabditis briggsae</td>\n",
       "      <td>C. briggsae</td>\n",
       "      <td>briggsae</td>\n",
       "      <td>6238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>Caenorhabditis</td>\n",
       "      <td>Caenorhabditis inopinata</td>\n",
       "      <td>C. inopinata</td>\n",
       "      <td>inopinata</td>\n",
       "      <td>1978547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>None</td>\n",
       "      <td>Caenorhabditis</td>\n",
       "      <td>Caenorhabditis japonica</td>\n",
       "      <td>C. japonica</td>\n",
       "      <td>japonica</td>\n",
       "      <td>281687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>None</td>\n",
       "      <td>Caenorhabditis</td>\n",
       "      <td>Caenorhabditis latens</td>\n",
       "      <td>C. latens</td>\n",
       "      <td>latens</td>\n",
       "      <td>1503980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>None</td>\n",
       "      <td>Caenorhabditis</td>\n",
       "      <td>Caenorhabditis nigoni</td>\n",
       "      <td>C. nigoni</td>\n",
       "      <td>nigoni</td>\n",
       "      <td>1611254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>None</td>\n",
       "      <td>Caenorhabditis</td>\n",
       "      <td>Caenorhabditis remanei</td>\n",
       "      <td>C. remanei</td>\n",
       "      <td>remanei</td>\n",
       "      <td>31234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>None</td>\n",
       "      <td>Caenorhabditis</td>\n",
       "      <td>Caenorhabditis sinica</td>\n",
       "      <td>C. sinica</td>\n",
       "      <td>sinica</td>\n",
       "      <td>1550068</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  commonName           genus                      name     shortName  \\\n",
       "0  roundworm  Caenorhabditis    Caenorhabditis elegans    C. elegans   \n",
       "1       None  Caenorhabditis    Caenorhabditis angaria    C. angaria   \n",
       "2       None  Caenorhabditis   Caenorhabditis brenneri   C. brenneri   \n",
       "3       None  Caenorhabditis   Caenorhabditis briggsae   C. briggsae   \n",
       "4       None  Caenorhabditis  Caenorhabditis inopinata  C. inopinata   \n",
       "5       None  Caenorhabditis   Caenorhabditis japonica   C. japonica   \n",
       "6       None  Caenorhabditis     Caenorhabditis latens     C. latens   \n",
       "7       None  Caenorhabditis     Caenorhabditis nigoni     C. nigoni   \n",
       "8       None  Caenorhabditis    Caenorhabditis remanei    C. remanei   \n",
       "9       None  Caenorhabditis     Caenorhabditis sinica     C. sinica   \n",
       "\n",
       "     species  taxonId  \n",
       "0    elegans     6239  \n",
       "1    angaria   860376  \n",
       "2   brenneri   135651  \n",
       "3   briggsae     6238  \n",
       "4  inopinata  1978547  \n",
       "5   japonica   281687  \n",
       "6     latens  1503980  \n",
       "7     nigoni  1611254  \n",
       "8    remanei    31234  \n",
       "9     sinica  1550068  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query(\"Organism\")\n",
    "query.select(\"commonName\", \"genus\", \"name\", \"shortName\", \"species\",\"taxonId\")\n",
    "\n",
    "query.add_constraint(\"genus\",\"=\",\"Caenorhabditis\") #Add a constraint to the query based on a column\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "organisms_data = pd.DataFrame(columns = [\"commonName\", \"genus\", \"name\", \"shortName\", \"species\",\"taxonId\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'commonName':row[0], 'genus':row[1], 'name':row[2], 'shortName':row[3], 'species':row[4], \n",
    "            'taxonId':row[5]}\n",
    "    organisms_data = organisms_data.append(info, ignore_index = True)\n",
    "organisms_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's query the WormMine database to extract some details of all genes for which data is available. Then we can add multiple constraints. It is not necessary that the constraints are related to the fields explicitly mentioned in the query. \n",
    "\n",
    "Here, we add a constraint based on the value of genus being Caenorhabditis and another based on the value of ontologyTerm being kinase activity. \n",
    "\n",
    "(Even though the genus field is used as a constraint, the query does not return the column of genus as it has not been called in the query.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>primaryIdentifier</th>\n",
       "      <th>ontologyAnnotations.id</th>\n",
       "      <th>ontologyAnnotations.qualifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>5000031</td>\n",
       "      <td>enables</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WBGene00000018</td>\n",
       "      <td>5000365</td>\n",
       "      <td>enables</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WBGene00000090</td>\n",
       "      <td>5003503</td>\n",
       "      <td>enables</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WBGene00000090</td>\n",
       "      <td>5003504</td>\n",
       "      <td>enables</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WBGene00000098</td>\n",
       "      <td>5003908</td>\n",
       "      <td>enables</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WBGene00000099</td>\n",
       "      <td>5004016</td>\n",
       "      <td>enables</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WBGene00000101</td>\n",
       "      <td>5004235</td>\n",
       "      <td>enables</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WBGene00000102</td>\n",
       "      <td>5004261</td>\n",
       "      <td>enables</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WBGene00000103</td>\n",
       "      <td>5004402</td>\n",
       "      <td>enables</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WBGene00000186</td>\n",
       "      <td>5006755</td>\n",
       "      <td>enables</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  primaryIdentifier ontologyAnnotations.id ontologyAnnotations.qualifier\n",
       "0    WBGene00000001                5000031                       enables\n",
       "1    WBGene00000018                5000365                       enables\n",
       "2    WBGene00000090                5003503                       enables\n",
       "3    WBGene00000090                5003504                       enables\n",
       "4    WBGene00000098                5003908                       enables\n",
       "5    WBGene00000099                5004016                       enables\n",
       "6    WBGene00000101                5004235                       enables\n",
       "7    WBGene00000102                5004261                       enables\n",
       "8    WBGene00000103                5004402                       enables\n",
       "9    WBGene00000186                5006755                       enables"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query(\"Gene\")\n",
    "query.select(\"primaryIdentifier\", \"ontologyAnnotations.id\", \"ontologyAnnotations.qualifier\")\n",
    "\n",
    "query.add_constraint(\"organism.genus\",\"=\",\"Caenorhabditis\") #Add a constraint to the query based on a column\n",
    "\n",
    "query.add_constraint(\"ontologyAnnotations.ontologyTerm.name\",\"=\",\"kinase activity\") #Add a second constraint\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "genes_data = pd.DataFrame(columns = [\"primaryIdentifier\", \"ontologyAnnotations.id\", \n",
    "                                     \"ontologyAnnotations.qualifier\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'primaryIdentifier':row[0], 'ontologyAnnotations.id':row[1], 'ontologyAnnotations.qualifier':row[2]}\n",
    "    genes_data = genes_data.append(info, ignore_index = True)\n",
    "genes_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's query the WormMine database to extract all homologue data that is available along with the gene primaryIdentifiers and symbols for this data. Like mentiones previously, we can add multiple constraints. But it is not necessary that the constraints are applied one on the other.\n",
    "\n",
    "We can use logic operators to apply constraints on the query outputs. & - AND and | - OR can be used in the regular sense.\n",
    "\n",
    "Here, we have 3 constraints based on - the value of genus being Caenorhabditis, the value of species being elegans, and type of homologue being orthologue.\n",
    "\n",
    "Then we apply the constraints on the query such that the output either has to have Caenorhabditis as genus AND elegans as species OR type of homologue as orthologue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Homologue.id</th>\n",
       "      <th>Homologue.type</th>\n",
       "      <th>gene.primaryIdentifier</th>\n",
       "      <th>gene.symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>39000008</td>\n",
       "      <td>orthologue</td>\n",
       "      <td>WBGene00004790</td>\n",
       "      <td>sgn-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39000012</td>\n",
       "      <td>orthologue</td>\n",
       "      <td>WBGene00004790</td>\n",
       "      <td>sgn-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>39000016</td>\n",
       "      <td>orthologue</td>\n",
       "      <td>WBGene00004790</td>\n",
       "      <td>sgn-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>39000021</td>\n",
       "      <td>orthologue</td>\n",
       "      <td>WBGene00004790</td>\n",
       "      <td>sgn-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>39000025</td>\n",
       "      <td>orthologue</td>\n",
       "      <td>WBGene00004790</td>\n",
       "      <td>sgn-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>39000029</td>\n",
       "      <td>orthologue</td>\n",
       "      <td>WBGene00004790</td>\n",
       "      <td>sgn-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>39000034</td>\n",
       "      <td>orthologue</td>\n",
       "      <td>WBGene00004790</td>\n",
       "      <td>sgn-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>39000060</td>\n",
       "      <td>orthologue</td>\n",
       "      <td>WBGene00008583</td>\n",
       "      <td>ugt-65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>39000064</td>\n",
       "      <td>orthologue</td>\n",
       "      <td>WBGene00009450</td>\n",
       "      <td>ugt-58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>39000068</td>\n",
       "      <td>orthologue</td>\n",
       "      <td>WBGene00011238</td>\n",
       "      <td>ugt-59</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Homologue.id Homologue.type gene.primaryIdentifier gene.symbol\n",
       "0     39000008     orthologue         WBGene00004790       sgn-1\n",
       "1     39000012     orthologue         WBGene00004790       sgn-1\n",
       "2     39000016     orthologue         WBGene00004790       sgn-1\n",
       "3     39000021     orthologue         WBGene00004790       sgn-1\n",
       "4     39000025     orthologue         WBGene00004790       sgn-1\n",
       "5     39000029     orthologue         WBGene00004790       sgn-1\n",
       "6     39000034     orthologue         WBGene00004790       sgn-1\n",
       "7     39000060     orthologue         WBGene00008583      ugt-65\n",
       "8     39000064     orthologue         WBGene00009450      ugt-58\n",
       "9     39000068     orthologue         WBGene00011238      ugt-59"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query(\"Homologue\")\n",
    "query.select('Homologue.id', 'Homologue.type', 'gene.primaryIdentifier', 'gene.symbol')\n",
    "\n",
    "query.add_constraint(\"gene.organism.genus\",\"=\",\"Caenorhabditis\") #Set up constraint A\n",
    "query.add_constraint(\"gene.organism.species\",\"=\",\"elegans\") #Set up constraint B\n",
    "query.add_constraint(\"type\",\"=\",\"orthologue\") #Set up constraint C\n",
    "\n",
    "query.set_logic(\"A & B & C\") #Logic operators can be used to set the different constraints on the query\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "homologue_data = pd.DataFrame(columns = [\"Homologue.id\", \"Homologue.type\", \"gene.primaryIdentifier\", \n",
    "                                         \"gene.symbol\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'Homologue.id':row[0], 'Homologue.type':row[1], 'gene.primaryIdentifier':row[2], 'gene.symbol':row[3]}\n",
    "    homologue_data = homologue_data.append(info, ignore_index = True)\n",
    "homologue_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Different types of constraints\n",
    "\n",
    "There are several kinds of constraints - Unary, Binary, Ternary, Multi-value, and List. We explore examples for all these constraint types."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Unary Constraints - \n",
    "Constraints that do not take any particular value but can be used to check if particular attirbute is absent or present.\n",
    "\n",
    "Types of Unary Constraints - IS Null and IS NOT Null\n",
    "\n",
    "Let's query the WormMine database to extract the automatedDescription, biotype, briefDescription, length, operon, primaryIdentifier, secondaryIdentifier, and symbol of all genes whose data is available, and then retain only those results where there is a valid secondaryIdentifier!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>automatedDescription</th>\n",
       "      <th>biotype</th>\n",
       "      <th>briefDescription</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>primaryIdentifier</th>\n",
       "      <th>secondaryIdentifier</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADP-ribosylation factor 4D is a member of the ...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>HGNC:656</td>\n",
       "      <td>ARL4D</td>\n",
       "      <td>Arl4d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADP-ribosylation factor-like 4A is a member of...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>HGNC:695</td>\n",
       "      <td>ARL4A</td>\n",
       "      <td>Arl4a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADP-ribosylation factor-like 4C is a member of...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>HGNC:698</td>\n",
       "      <td>ARL4C</td>\n",
       "      <td>Arl4c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADP-ribosyltransferase catalyzes the ADP-ribos...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>HGNC:723</td>\n",
       "      <td>ART1</td>\n",
       "      <td>Art1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AIM2 is a member of the IFI20X /IFI16 family. ...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>HGNC:357</td>\n",
       "      <td>AIM2</td>\n",
       "      <td>Aim2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ASSOCIATED WITH Atrioventricular Septal Defect 4</td>\n",
       "      <td>SO:0001267</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>HGNC:50403</td>\n",
       "      <td>SNORA99</td>\n",
       "      <td>SNORA99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ASSOCIATED WITH Bardet-Biedl Syndrome, 21; con...</td>\n",
       "      <td>SO:0001263</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>HGNC:50444</td>\n",
       "      <td>C8orf37-AS1</td>\n",
       "      <td>C8orf37-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ASSOCIATED WITH Chromosome 13q Deletion Syndro...</td>\n",
       "      <td>SO:0001263</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>HGNC:50277</td>\n",
       "      <td>LMO7-AS1</td>\n",
       "      <td>LMO7-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ASSOCIATED WITH Chromosome 13q Deletion Syndro...</td>\n",
       "      <td>SO:0001263</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>HGNC:50496</td>\n",
       "      <td>DLEU1-AS1</td>\n",
       "      <td>DLEU1-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ASSOCIATED WITH Experimental Liver Cirrhosis; ...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>HGNC:1310</td>\n",
       "      <td>TMEM184B</td>\n",
       "      <td>Tmem184b</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                automatedDescription     biotype  \\\n",
       "0  ADP-ribosylation factor 4D is a member of the ...  SO:0001217   \n",
       "1  ADP-ribosylation factor-like 4A is a member of...  SO:0001217   \n",
       "2  ADP-ribosylation factor-like 4C is a member of...  SO:0001217   \n",
       "3  ADP-ribosyltransferase catalyzes the ADP-ribos...  SO:0001217   \n",
       "4  AIM2 is a member of the IFI20X /IFI16 family. ...  SO:0001217   \n",
       "5   ASSOCIATED WITH Atrioventricular Septal Defect 4  SO:0001267   \n",
       "6  ASSOCIATED WITH Bardet-Biedl Syndrome, 21; con...  SO:0001263   \n",
       "7  ASSOCIATED WITH Chromosome 13q Deletion Syndro...  SO:0001263   \n",
       "8  ASSOCIATED WITH Chromosome 13q Deletion Syndro...  SO:0001263   \n",
       "9  ASSOCIATED WITH Experimental Liver Cirrhosis; ...  SO:0001217   \n",
       "\n",
       "  briefDescription length operon primaryIdentifier secondaryIdentifier  \\\n",
       "0             None   None   None          HGNC:656               ARL4D   \n",
       "1             None   None   None          HGNC:695               ARL4A   \n",
       "2             None   None   None          HGNC:698               ARL4C   \n",
       "3             None   None   None          HGNC:723                ART1   \n",
       "4             None   None   None          HGNC:357                AIM2   \n",
       "5             None   None   None        HGNC:50403             SNORA99   \n",
       "6             None   None   None        HGNC:50444         C8orf37-AS1   \n",
       "7             None   None   None        HGNC:50277            LMO7-AS1   \n",
       "8             None   None   None        HGNC:50496           DLEU1-AS1   \n",
       "9             None   None   None         HGNC:1310            TMEM184B   \n",
       "\n",
       "        symbol  \n",
       "0        Arl4d  \n",
       "1        Arl4a  \n",
       "2        Arl4c  \n",
       "3         Art1  \n",
       "4         Aim2  \n",
       "5      SNORA99  \n",
       "6  C8orf37-AS1  \n",
       "7     LMO7-AS1  \n",
       "8    DLEU1-AS1  \n",
       "9     Tmem184b  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query(\"Gene\")\n",
    "query.select(\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \"primaryIdentifier\", \n",
    "             \"secondaryIdentifier\", \"symbol\")\n",
    "\n",
    "query.add_constraint(\"secondaryIdentifier\",\"IS NOT NULL\") #Unary Constraint\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "genes_data = pd.DataFrame(columns = [\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \n",
    "                                     \"primaryIdentifier\", \"secondaryIdentifier\", \"symbol\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'automatedDescription':row[0], 'biotype':row[1], 'briefDescription':row[2], 'length':row[3], \n",
    "            'operon':row[4], 'primaryIdentifier':row[5], 'secondaryIdentifier':row[6], 'symbol':row[7]}\n",
    "    genes_data = genes_data.append(info, ignore_index = True)\n",
    "genes_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Binary Constraints - \n",
    "Constraints that can take a particular attribute which can then be compared them to a specified value.\n",
    "\n",
    "Types of Binary Constraints - `=`, `<=`, `>=`, `<`, `>`, `!=`.\n",
    "\n",
    "Let's query the WormMine database to extract the automatedDescription, biotype, briefDescription, length, operon, primaryIdentifier, secondaryIdentifier, and symbol of all genes whose data is available, and then retain only those results where the value of the length is greater than or equal to 12000!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>automatedDescription</th>\n",
       "      <th>biotype</th>\n",
       "      <th>briefDescription</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>primaryIdentifier</th>\n",
       "      <th>secondaryIdentifier</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Enables ATP-dependent microtubule motor activi...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>unc-104 encodes a kinesin-like motor protein h...</td>\n",
       "      <td>15639</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00006831</td>\n",
       "      <td>C52E12.2</td>\n",
       "      <td>unc-104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Enables ATPase-coupled intramembrane lipid tra...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>tat-1 encodes one of six C. elegans subfamily ...</td>\n",
       "      <td>22818</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00013034</td>\n",
       "      <td>Y49E10.11</td>\n",
       "      <td>tat-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Enables DEAD/H-box RNA helicase binding activi...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>zyx-1 encodes a LIM domain protein and is most...</td>\n",
       "      <td>14604</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00006999</td>\n",
       "      <td>F42G4.3</td>\n",
       "      <td>zyx-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Enables DNA-binding transcription factor activ...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>sbp-1 encodes a basic helix-loop-helix (bHLH) ...</td>\n",
       "      <td>14200</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00004735</td>\n",
       "      <td>Y47D3B.7</td>\n",
       "      <td>sbp-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Enables DNA-binding transcription factor activ...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>vab-3 encodes a homeodomain protein that is th...</td>\n",
       "      <td>19691</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00006870</td>\n",
       "      <td>F14F3.1</td>\n",
       "      <td>vab-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Enables G protein-coupled GABA receptor activi...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>gbb-2 encodes a subunit of the GABAB receptor ...</td>\n",
       "      <td>17820</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00022675</td>\n",
       "      <td>ZK180.1</td>\n",
       "      <td>gbb-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Enables G protein-coupled acetylcholine recept...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>gar-3 encodes a G protein-linked muscarinic ac...</td>\n",
       "      <td>14901</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00001519</td>\n",
       "      <td>Y40H4A.1</td>\n",
       "      <td>gar-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Enables G-protein alpha-subunit binding activi...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>ags-3 encodes a protein containing N-terminal ...</td>\n",
       "      <td>13877</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00000092</td>\n",
       "      <td>F32A6.4</td>\n",
       "      <td>ags-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Enables GABA receptor binding activity. Is inv...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>12622</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00001490</td>\n",
       "      <td>H05G16.1</td>\n",
       "      <td>frm-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Enables GTPase activating protein binding acti...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>mpz-1 encodes a multi-PDZ domain scaffold prot...</td>\n",
       "      <td>39492</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00003404</td>\n",
       "      <td>C52A11.4</td>\n",
       "      <td>mpz-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                automatedDescription     biotype  \\\n",
       "0  Enables ATP-dependent microtubule motor activi...  SO:0001217   \n",
       "1  Enables ATPase-coupled intramembrane lipid tra...  SO:0001217   \n",
       "2  Enables DEAD/H-box RNA helicase binding activi...  SO:0001217   \n",
       "3  Enables DNA-binding transcription factor activ...  SO:0001217   \n",
       "4  Enables DNA-binding transcription factor activ...  SO:0001217   \n",
       "5  Enables G protein-coupled GABA receptor activi...  SO:0001217   \n",
       "6  Enables G protein-coupled acetylcholine recept...  SO:0001217   \n",
       "7  Enables G-protein alpha-subunit binding activi...  SO:0001217   \n",
       "8  Enables GABA receptor binding activity. Is inv...  SO:0001217   \n",
       "9  Enables GTPase activating protein binding acti...  SO:0001217   \n",
       "\n",
       "                                    briefDescription length operon  \\\n",
       "0  unc-104 encodes a kinesin-like motor protein h...  15639   None   \n",
       "1  tat-1 encodes one of six C. elegans subfamily ...  22818   None   \n",
       "2  zyx-1 encodes a LIM domain protein and is most...  14604   None   \n",
       "3  sbp-1 encodes a basic helix-loop-helix (bHLH) ...  14200   None   \n",
       "4  vab-3 encodes a homeodomain protein that is th...  19691   None   \n",
       "5  gbb-2 encodes a subunit of the GABAB receptor ...  17820   None   \n",
       "6  gar-3 encodes a G protein-linked muscarinic ac...  14901   None   \n",
       "7  ags-3 encodes a protein containing N-terminal ...  13877   None   \n",
       "8                                               None  12622   None   \n",
       "9  mpz-1 encodes a multi-PDZ domain scaffold prot...  39492   None   \n",
       "\n",
       "  primaryIdentifier secondaryIdentifier   symbol  \n",
       "0    WBGene00006831            C52E12.2  unc-104  \n",
       "1    WBGene00013034           Y49E10.11    tat-1  \n",
       "2    WBGene00006999             F42G4.3    zyx-1  \n",
       "3    WBGene00004735            Y47D3B.7    sbp-1  \n",
       "4    WBGene00006870             F14F3.1    vab-3  \n",
       "5    WBGene00022675             ZK180.1    gbb-2  \n",
       "6    WBGene00001519            Y40H4A.1    gar-3  \n",
       "7    WBGene00000092             F32A6.4    ags-3  \n",
       "8    WBGene00001490            H05G16.1    frm-3  \n",
       "9    WBGene00003404            C52A11.4    mpz-1  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query(\"Gene\")\n",
    "query.select(\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \"primaryIdentifier\", \n",
    "             \"secondaryIdentifier\", \"symbol\")\n",
    "\n",
    "query.add_constraint(\"length\",\">=\",\"12000\") #Binary Constraint\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "genes_data = pd.DataFrame(columns = [\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \n",
    "                                     \"primaryIdentifier\", \"secondaryIdentifier\", \"symbol\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'automatedDescription':row[0], 'biotype':row[1], 'briefDescription':row[2], 'length':row[3], \n",
    "            'operon':row[4], 'primaryIdentifier':row[5], 'secondaryIdentifier':row[6], 'symbol':row[7]}\n",
    "    genes_data = genes_data.append(info, ignore_index = True)\n",
    "genes_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Ternary constraints - \n",
    "Constraints that have one required value and one optional value.\n",
    "\n",
    "Types of Ternary Constraints -LOOKUP (this operator searches through all the fields in a particular class for the value specified by the user)\n",
    "\n",
    "Let's query the WormMine database to extract the automatedDescription, biotype, briefDescription, length, operon, primaryIdentifier, secondaryIdentifier, and symbol of all genes whose data is available, and then retain only those results where there is a mention of 'GABA' in any field!\n",
    "\n",
    "The extra_value parameter can be used to limit the search to the type of object (for example, organism in genes), here C. elegans."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>automatedDescription</th>\n",
       "      <th>biotype</th>\n",
       "      <th>briefDescription</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>primaryIdentifier</th>\n",
       "      <th>secondaryIdentifier</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Enables several functions, including DNA-bindi...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>hlh-2 encodes a Class I basic helix-loop-helix...</td>\n",
       "      <td>3133</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00001949</td>\n",
       "      <td>M05B5.5</td>\n",
       "      <td>hlh-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                automatedDescription     biotype  \\\n",
       "0  Enables several functions, including DNA-bindi...  SO:0001217   \n",
       "\n",
       "                                    briefDescription length operon  \\\n",
       "0  hlh-2 encodes a Class I basic helix-loop-helix...   3133   None   \n",
       "\n",
       "  primaryIdentifier secondaryIdentifier symbol  \n",
       "0    WBGene00001949             M05B5.5  hlh-2  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query(\"Gene\")\n",
    "query.select(\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \"primaryIdentifier\", \n",
    "             \"secondaryIdentifier\", \"symbol\")\n",
    "\n",
    "query.add_constraint(\"Gene\", \"LOOKUP\", \"hlh-2\", extra_value='C. elegans') #Ternary Constraint\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "genes_data = pd.DataFrame(columns = [\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \n",
    "                                     \"primaryIdentifier\", \"secondaryIdentifier\", \"symbol\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'automatedDescription':row[0], 'biotype':row[1], 'briefDescription':row[2], 'length':row[3], \n",
    "            'operon':row[4], 'primaryIdentifier':row[5], 'secondaryIdentifier':row[6], 'symbol':row[7]}\n",
    "    genes_data = genes_data.append(info, ignore_index = True)\n",
    "genes_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Multi-value Constraints - \n",
    "Constraints that can take multiple values.\n",
    "\n",
    "Types of Multi-value Constraints -ONE OF and NONE OF\n",
    "\n",
    "Let's query the WormMine database to extract the automatedDescription, biotype, briefDescription, length, operon, primaryIdentifier, secondaryIdentifier, and symbol of all genes whose data is available, and then retain only those results where the gene symbol is one of hlh-2, unc-26, gar-3, or gbb-2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>automatedDescription</th>\n",
       "      <th>biotype</th>\n",
       "      <th>briefDescription</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>primaryIdentifier</th>\n",
       "      <th>secondaryIdentifier</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Enables G protein-coupled GABA receptor activi...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>gbb-2 encodes a subunit of the GABAB receptor ...</td>\n",
       "      <td>17820</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00022675</td>\n",
       "      <td>ZK180.1</td>\n",
       "      <td>gbb-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Enables G protein-coupled acetylcholine recept...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>gar-3 encodes a G protein-linked muscarinic ac...</td>\n",
       "      <td>14901</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00001519</td>\n",
       "      <td>Y40H4A.1</td>\n",
       "      <td>gar-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Enables several functions, including DNA-bindi...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>hlh-2 encodes a Class I basic helix-loop-helix...</td>\n",
       "      <td>3133</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00001949</td>\n",
       "      <td>M05B5.5</td>\n",
       "      <td>hlh-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Is predicted to enable phosphatidylinositol-3,...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>unc-26 encodes synaptojanin, a polyphosphoinos...</td>\n",
       "      <td>10962</td>\n",
       "      <td>CEOP4539</td>\n",
       "      <td>WBGene00006763</td>\n",
       "      <td>JC8.10</td>\n",
       "      <td>unc-26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                automatedDescription     biotype  \\\n",
       "0  Enables G protein-coupled GABA receptor activi...  SO:0001217   \n",
       "1  Enables G protein-coupled acetylcholine recept...  SO:0001217   \n",
       "2  Enables several functions, including DNA-bindi...  SO:0001217   \n",
       "3  Is predicted to enable phosphatidylinositol-3,...  SO:0001217   \n",
       "\n",
       "                                    briefDescription length    operon  \\\n",
       "0  gbb-2 encodes a subunit of the GABAB receptor ...  17820      None   \n",
       "1  gar-3 encodes a G protein-linked muscarinic ac...  14901      None   \n",
       "2  hlh-2 encodes a Class I basic helix-loop-helix...   3133      None   \n",
       "3  unc-26 encodes synaptojanin, a polyphosphoinos...  10962  CEOP4539   \n",
       "\n",
       "  primaryIdentifier secondaryIdentifier  symbol  \n",
       "0    WBGene00022675             ZK180.1   gbb-2  \n",
       "1    WBGene00001519            Y40H4A.1   gar-3  \n",
       "2    WBGene00001949             M05B5.5   hlh-2  \n",
       "3    WBGene00006763              JC8.10  unc-26  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query(\"Gene\")\n",
    "query.select(\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \"primaryIdentifier\", \n",
    "             \"secondaryIdentifier\", \"symbol\")\n",
    "\n",
    "query.add_constraint(\"symbol\",\"ONE OF\",['hlh-2','unc-26', 'gar-3', 'gbb-2']) #Multi-value Constraint\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "genes_data = pd.DataFrame(columns = [\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \n",
    "                                     \"primaryIdentifier\", \"secondaryIdentifier\", \"symbol\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'automatedDescription':row[0], 'biotype':row[1], 'briefDescription':row[2], 'length':row[3], \n",
    "            'operon':row[4], 'primaryIdentifier':row[5], 'secondaryIdentifier':row[6], 'symbol':row[7]}\n",
    "    genes_data = genes_data.append(info, ignore_index = True)\n",
    "genes_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### List constraints - \n",
    "Constraints that contain a list of values.\n",
    "\n",
    "Types of List Constraints -IN or NOT IN \n",
    "\n",
    "Let's query the WormMine database to extract the automatedDescription, biotype, briefDescription, length, operon, primaryIdentifier, secondaryIdentifier, and symbol of all genes whose data is available, and then retain only those results where the Gene is in the publicly available list C. elegans transcription factor genes present on WormMine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>automatedDescription</th>\n",
       "      <th>biotype</th>\n",
       "      <th>briefDescription</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>primaryIdentifier</th>\n",
       "      <th>secondaryIdentifier</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acts upstream of or within positive regulation...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>The tlp-1 gene encodes a C2H2-type zinc finger...</td>\n",
       "      <td>2875</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00006580</td>\n",
       "      <td>T23G4.1</td>\n",
       "      <td>tlp-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Contributes to RNA polymerase II activity. Is ...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>ama-1 encodes the large subunit of RNA polymer...</td>\n",
       "      <td>10100</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00000123</td>\n",
       "      <td>F36A4.7</td>\n",
       "      <td>ama-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Contributes to RNA polymerase II transcription...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>mab-5 encodes a homeodomain transcription fact...</td>\n",
       "      <td>6620</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00003102</td>\n",
       "      <td>C08C3.3</td>\n",
       "      <td>mab-5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Contributes to sequence-specific DNA binding a...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>cky-1 encodes a member of the basic helix-loop...</td>\n",
       "      <td>4503</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00000521</td>\n",
       "      <td>C15C8.2</td>\n",
       "      <td>cky-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Enables DEAD/H-box RNA helicase binding activi...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>zyx-1 encodes a LIM domain protein and is most...</td>\n",
       "      <td>14604</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00006999</td>\n",
       "      <td>F42G4.3</td>\n",
       "      <td>zyx-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Enables DEAD/H-box RNA helicase binding activi...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>nhl-2 encodes one of five C. elegans proteins ...</td>\n",
       "      <td>5428</td>\n",
       "      <td>CEOP3232</td>\n",
       "      <td>WBGene00003598</td>\n",
       "      <td>F26F4.7</td>\n",
       "      <td>nhl-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Enables DNA binding activity and RNA polymeras...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>ahr-1 encodes an aryl hydrocarbon receptor (li...</td>\n",
       "      <td>8583</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00000096</td>\n",
       "      <td>C41G7.5</td>\n",
       "      <td>ahr-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Enables DNA binding activity, bending and doub...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>ceh-37 encodes one of three C. elegans protein...</td>\n",
       "      <td>10629</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00000458</td>\n",
       "      <td>C37E2.5</td>\n",
       "      <td>ceh-37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Enables DNA binding activity; DNA-binding tran...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>aha-1 encodes an ortholog of human aryl-hydroc...</td>\n",
       "      <td>3073</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00000095</td>\n",
       "      <td>C25A1.11</td>\n",
       "      <td>aha-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Enables DNA-binding transcription activator ac...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>ceh-17 encodes a phox-2-like homeodomain prote...</td>\n",
       "      <td>1263</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00000440</td>\n",
       "      <td>D1007.1</td>\n",
       "      <td>ceh-17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                automatedDescription     biotype  \\\n",
       "0  Acts upstream of or within positive regulation...  SO:0001217   \n",
       "1  Contributes to RNA polymerase II activity. Is ...  SO:0001217   \n",
       "2  Contributes to RNA polymerase II transcription...  SO:0001217   \n",
       "3  Contributes to sequence-specific DNA binding a...  SO:0001217   \n",
       "4  Enables DEAD/H-box RNA helicase binding activi...  SO:0001217   \n",
       "5  Enables DEAD/H-box RNA helicase binding activi...  SO:0001217   \n",
       "6  Enables DNA binding activity and RNA polymeras...  SO:0001217   \n",
       "7  Enables DNA binding activity, bending and doub...  SO:0001217   \n",
       "8  Enables DNA binding activity; DNA-binding tran...  SO:0001217   \n",
       "9  Enables DNA-binding transcription activator ac...  SO:0001217   \n",
       "\n",
       "                                    briefDescription length    operon  \\\n",
       "0  The tlp-1 gene encodes a C2H2-type zinc finger...   2875      None   \n",
       "1  ama-1 encodes the large subunit of RNA polymer...  10100      None   \n",
       "2  mab-5 encodes a homeodomain transcription fact...   6620      None   \n",
       "3  cky-1 encodes a member of the basic helix-loop...   4503      None   \n",
       "4  zyx-1 encodes a LIM domain protein and is most...  14604      None   \n",
       "5  nhl-2 encodes one of five C. elegans proteins ...   5428  CEOP3232   \n",
       "6  ahr-1 encodes an aryl hydrocarbon receptor (li...   8583      None   \n",
       "7  ceh-37 encodes one of three C. elegans protein...  10629      None   \n",
       "8  aha-1 encodes an ortholog of human aryl-hydroc...   3073      None   \n",
       "9  ceh-17 encodes a phox-2-like homeodomain prote...   1263      None   \n",
       "\n",
       "  primaryIdentifier secondaryIdentifier  symbol  \n",
       "0    WBGene00006580             T23G4.1   tlp-1  \n",
       "1    WBGene00000123             F36A4.7   ama-1  \n",
       "2    WBGene00003102             C08C3.3   mab-5  \n",
       "3    WBGene00000521             C15C8.2   cky-1  \n",
       "4    WBGene00006999             F42G4.3   zyx-1  \n",
       "5    WBGene00003598             F26F4.7   nhl-2  \n",
       "6    WBGene00000096             C41G7.5   ahr-1  \n",
       "7    WBGene00000458             C37E2.5  ceh-37  \n",
       "8    WBGene00000095            C25A1.11   aha-1  \n",
       "9    WBGene00000440             D1007.1  ceh-17  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query(\"Gene\")\n",
    "query.select(\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \"primaryIdentifier\", \n",
    "             \"secondaryIdentifier\", \"symbol\")\n",
    "\n",
    "query.add_constraint(\"Gene\",\"IN\",\"C. elegans transcription factor genes\") #List Constraint\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "genes_data = pd.DataFrame(columns = [\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \n",
    "                                     \"primaryIdentifier\", \"secondaryIdentifier\", \"symbol\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'automatedDescription':row[0], 'biotype':row[1], 'briefDescription':row[2], 'length':row[3], \n",
    "            'operon':row[4], 'primaryIdentifier':row[5], 'secondaryIdentifier':row[6], 'symbol':row[7]}\n",
    "    genes_data = genes_data.append(info, ignore_index = True)\n",
    "genes_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Sub-Class constraints - \n",
    "Constraints that allow a user to specify a sub-class of a class to constrain a path to\n",
    "\n",
    "Let's query the WormMine database to extract the automatedDescription, biotype, briefDescription, length, operon, primaryIdentifier, secondaryIdentifier, and symbol of all genes whose data is available, and then constrain our results to only those items of the sub class GOAnnotation of ontologyAnnotations class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>automatedDescription</th>\n",
       "      <th>biotype</th>\n",
       "      <th>briefDescription</th>\n",
       "      <th>length</th>\n",
       "      <th>operon</th>\n",
       "      <th>primaryIdentifier</th>\n",
       "      <th>secondaryIdentifier</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acts upstream of or within IRE1-mediated unfol...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>nlp-28 encodes a neuropeptide-like protein.</td>\n",
       "      <td>417</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00003766</td>\n",
       "      <td>B0213.3</td>\n",
       "      <td>nlp-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acts upstream of or within IRE1-mediated unfol...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>468</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00007992</td>\n",
       "      <td>C37A5.8</td>\n",
       "      <td>fipr-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acts upstream of or within IRE1-mediated unfol...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>2395</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00009239</td>\n",
       "      <td>F28H7.6</td>\n",
       "      <td>irld-6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acts upstream of or within IRE1-mediated unfol...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>961</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00018181</td>\n",
       "      <td>F38E1.9</td>\n",
       "      <td>mpdu-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acts upstream of or within IRE1-mediated unfol...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>3594</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00016469</td>\n",
       "      <td>C36B7.6</td>\n",
       "      <td>C36B7.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Acts upstream of or within IRE1-mediated unfol...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>1699</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00009112</td>\n",
       "      <td>F25D7.2</td>\n",
       "      <td>tag-353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Acts upstream of or within IRE1-mediated unfol...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>979</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00017705</td>\n",
       "      <td>F22E5.6</td>\n",
       "      <td>F22E5.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Acts upstream of or within IRE1-mediated unfol...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>1097</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00022570</td>\n",
       "      <td>ZC239.12</td>\n",
       "      <td>sdz-35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Acts upstream of or within IRE1-mediated unfol...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>1326</td>\n",
       "      <td>CEOPX072</td>\n",
       "      <td>WBGene00020486</td>\n",
       "      <td>T13C5.6</td>\n",
       "      <td>T13C5.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Acts upstream of or within IRE1-mediated unfol...</td>\n",
       "      <td>SO:0001217</td>\n",
       "      <td>None</td>\n",
       "      <td>2275</td>\n",
       "      <td>None</td>\n",
       "      <td>WBGene00010496</td>\n",
       "      <td>K02B12.3</td>\n",
       "      <td>sec-12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                automatedDescription     biotype  \\\n",
       "0  Acts upstream of or within IRE1-mediated unfol...  SO:0001217   \n",
       "1  Acts upstream of or within IRE1-mediated unfol...  SO:0001217   \n",
       "2  Acts upstream of or within IRE1-mediated unfol...  SO:0001217   \n",
       "3  Acts upstream of or within IRE1-mediated unfol...  SO:0001217   \n",
       "4  Acts upstream of or within IRE1-mediated unfol...  SO:0001217   \n",
       "5  Acts upstream of or within IRE1-mediated unfol...  SO:0001217   \n",
       "6  Acts upstream of or within IRE1-mediated unfol...  SO:0001217   \n",
       "7  Acts upstream of or within IRE1-mediated unfol...  SO:0001217   \n",
       "8  Acts upstream of or within IRE1-mediated unfol...  SO:0001217   \n",
       "9  Acts upstream of or within IRE1-mediated unfol...  SO:0001217   \n",
       "\n",
       "                              briefDescription length    operon  \\\n",
       "0  nlp-28 encodes a neuropeptide-like protein.    417      None   \n",
       "1                                         None    468      None   \n",
       "2                                         None   2395      None   \n",
       "3                                         None    961      None   \n",
       "4                                         None   3594      None   \n",
       "5                                         None   1699      None   \n",
       "6                                         None    979      None   \n",
       "7                                         None   1097      None   \n",
       "8                                         None   1326  CEOPX072   \n",
       "9                                         None   2275      None   \n",
       "\n",
       "  primaryIdentifier secondaryIdentifier   symbol  \n",
       "0    WBGene00003766             B0213.3   nlp-28  \n",
       "1    WBGene00007992             C37A5.8  fipr-24  \n",
       "2    WBGene00009239             F28H7.6   irld-6  \n",
       "3    WBGene00018181             F38E1.9   mpdu-1  \n",
       "4    WBGene00016469             C36B7.6  C36B7.6  \n",
       "5    WBGene00009112             F25D7.2  tag-353  \n",
       "6    WBGene00017705             F22E5.6  F22E5.6  \n",
       "7    WBGene00022570            ZC239.12   sdz-35  \n",
       "8    WBGene00020486             T13C5.6  T13C5.6  \n",
       "9    WBGene00010496            K02B12.3   sec-12  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query(\"Gene\")\n",
    "query.select(\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \"primaryIdentifier\", \n",
    "             \"secondaryIdentifier\", \"symbol\")\n",
    "\n",
    "query.add_constraint(\"ontologyAnnotations\",\"GOAnnotation\") #Sub-Class constraint\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "genes_data = pd.DataFrame(columns = [\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \n",
    "                                     \"primaryIdentifier\", \"secondaryIdentifier\", \"symbol\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'automatedDescription':row[0], 'biotype':row[1], 'briefDescription':row[2], 'length':row[3], \n",
    "            'operon':row[4], 'primaryIdentifier':row[5], 'secondaryIdentifier':row[6], 'symbol':row[7]}\n",
    "    genes_data = genes_data.append(info, ignore_index = True)\n",
    "genes_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Loop Constraints - \n",
    "Constraints that assert that two paths refer to the same object\n",
    "\n",
    "Types of Loop Constraints - IS or IS NOT\n",
    "\n",
    "Let's query the WormMine database to extract the automatedDescription, biotype, briefDescription, length, operon, primaryIdentifier, secondaryIdentifier, and symbol of all genes whose data is available, constrain our results using a list constraint and then, a loop constraint."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>homologues.gene.primaryIdentifier</th>\n",
       "      <th>homologues.homologue.primaryIdentifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WBGene00006580</td>\n",
       "      <td>FBgn0004858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WBGene00006580</td>\n",
       "      <td>FBgn0005771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WBGene00006580</td>\n",
       "      <td>HGNC:23589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WBGene00006580</td>\n",
       "      <td>HGNC:25883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WBGene00006580</td>\n",
       "      <td>MGI:1353644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WBGene00006580</td>\n",
       "      <td>MGI:2662729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WBGene00006580</td>\n",
       "      <td>RGD:1312018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WBGene00006580</td>\n",
       "      <td>RGD:1588249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WBGene00006580</td>\n",
       "      <td>ZDB-GENE-010717-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WBGene00006580</td>\n",
       "      <td>ZDB-GENE-031113-5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  homologues.gene.primaryIdentifier homologues.homologue.primaryIdentifier\n",
       "0                    WBGene00006580                            FBgn0004858\n",
       "1                    WBGene00006580                            FBgn0005771\n",
       "2                    WBGene00006580                             HGNC:23589\n",
       "3                    WBGene00006580                             HGNC:25883\n",
       "4                    WBGene00006580                            MGI:1353644\n",
       "5                    WBGene00006580                            MGI:2662729\n",
       "6                    WBGene00006580                            RGD:1312018\n",
       "7                    WBGene00006580                            RGD:1588249\n",
       "8                    WBGene00006580                      ZDB-GENE-010717-1\n",
       "9                    WBGene00006580                      ZDB-GENE-031113-5"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query(\"Gene\")\n",
    "query.select(\"automatedDescription\", \"biotype\", \"briefDescription\", \"length\", \"operon\", \"primaryIdentifier\", \n",
    "             \"secondaryIdentifier\", \"symbol\")\n",
    "\n",
    "query.add_view(\"homologues.gene.primaryIdentifier\",\"homologues.homologue.primaryIdentifier\") #Add more fields\n",
    "\n",
    "query.add_constraint(\"Gene\", \"IN\", \"C. elegans transcription factor genes\", code = \"A\") #List constraint\n",
    "\n",
    "query.add_constraint(\"homologues.homologue\", \"IS NOT\", \"Gene\", code = \"B\") #Loop Constraint\n",
    "\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "genes_data = pd.DataFrame(columns = [\"homologues.gene.primaryIdentifier\", \"homologues.homologue.primaryIdentifier\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'homologues.gene.primaryIdentifier':row['homologues.gene.primaryIdentifier'], \n",
    "            'homologues.homologue.primaryIdentifier':row['homologues.homologue.primaryIdentifier']}\n",
    "    genes_data = genes_data.append(info, ignore_index = True)\n",
    "genes_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Range Constraints - \n",
    "Constraints that test whether a value lies relative to a set of ranges or not\n",
    "\n",
    "Types of Range Constraints - OVERLAPS, DOES NOT OVERLAP, WITHIN, OUTSIDE, CONTAINS and DOES NOT CONTAIN\n",
    "\n",
    "Let's query the WormMine database to extract the details on organism name and chromosome location of all sequences whose data is available, and then constrain our results based on if the chromosome location overlaps our specified range of I:1..4000."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Organism</th>\n",
       "      <th>Chromosome identifier</th>\n",
       "      <th>Chromosome Start Location</th>\n",
       "      <th>Chromosome End Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C. elegans</td>\n",
       "      <td>I</td>\n",
       "      <td>1</td>\n",
       "      <td>3746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C. elegans</td>\n",
       "      <td>I</td>\n",
       "      <td>3747</td>\n",
       "      <td>3909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C. elegans</td>\n",
       "      <td>I</td>\n",
       "      <td>3747</td>\n",
       "      <td>3909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C. elegans</td>\n",
       "      <td>I</td>\n",
       "      <td>3747</td>\n",
       "      <td>3909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C. elegans</td>\n",
       "      <td>I</td>\n",
       "      <td>3910</td>\n",
       "      <td>4115</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Organism Chromosome identifier Chromosome Start Location  \\\n",
       "0  C. elegans                     I                         1   \n",
       "1  C. elegans                     I                      3747   \n",
       "2  C. elegans                     I                      3747   \n",
       "3  C. elegans                     I                      3747   \n",
       "4  C. elegans                     I                      3910   \n",
       "\n",
       "  Chromosome End Location  \n",
       "0                    3746  \n",
       "1                    3909  \n",
       "2                    3909  \n",
       "3                    3909  \n",
       "4                    4115  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query()\n",
    "\n",
    "query.add_view(\"SequenceFeature.organism.shortName\", \n",
    "               \"SequenceFeature.chromosomeLocation.locatedOn.primaryIdentifier\", \n",
    "               \"SequenceFeature.chromosomeLocation.start\", \"SequenceFeature.chromosomeLocation.end\" ) #Add fields\n",
    "\n",
    "query.add_constraint(\"chromosomeLocation\", \"OVERLAPS\", [\"I:1..4000\"]) #Range constraint\n",
    "\n",
    "\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "genes_data = pd.DataFrame(columns = [\"Organism\", \"Chromosome identifier\", \"Chromosome Start Location\", \n",
    "                                     \"Chromosome End Location\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'Organism':row[0], 'Chromosome identifier':row[1], 'Chromosome Start Location':row[2], \n",
    "            'Chromosome End Location':row[3]}\n",
    "    genes_data = genes_data.append(info, ignore_index = True)\n",
    "genes_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Some query examples and exploring some other functionalities"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's query the WormMine database to extract the primaryIdentifier, symbol, of all genes whose data is available and connect that to the name and identifiers of the ontology terms. We then add a constraint that the homologue type of the results should be orthologue:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>primaryIdentifier</th>\n",
       "      <th>symbol</th>\n",
       "      <th>Ontology Name</th>\n",
       "      <th>Ontology Identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>1-phosphatidylinositol-3-kinase regulator acti...</td>\n",
       "      <td>GO:0046935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>dauer larval development</td>\n",
       "      <td>GO:0040024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>determination of adult lifespan</td>\n",
       "      <td>GO:0008340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>insulin receptor signaling pathway</td>\n",
       "      <td>GO:0008286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>kinase activity</td>\n",
       "      <td>GO:0016301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>phosphatidylinositol 3-kinase complex</td>\n",
       "      <td>GO:0005942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>phosphatidylinositol phosphorylation</td>\n",
       "      <td>GO:0046854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>phosphorylation</td>\n",
       "      <td>GO:0016310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>protein kinase binding</td>\n",
       "      <td>GO:0019901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>regulation of phosphatidylinositol 3-kinase ac...</td>\n",
       "      <td>GO:0043551</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  primaryIdentifier symbol                                      Ontology Name  \\\n",
       "0    WBGene00000001  aap-1  1-phosphatidylinositol-3-kinase regulator acti...   \n",
       "1    WBGene00000001  aap-1                           dauer larval development   \n",
       "2    WBGene00000001  aap-1                    determination of adult lifespan   \n",
       "3    WBGene00000001  aap-1                 insulin receptor signaling pathway   \n",
       "4    WBGene00000001  aap-1                                    kinase activity   \n",
       "5    WBGene00000001  aap-1              phosphatidylinositol 3-kinase complex   \n",
       "6    WBGene00000001  aap-1               phosphatidylinositol phosphorylation   \n",
       "7    WBGene00000001  aap-1                                    phosphorylation   \n",
       "8    WBGene00000001  aap-1                             protein kinase binding   \n",
       "9    WBGene00000001  aap-1  regulation of phosphatidylinositol 3-kinase ac...   \n",
       "\n",
       "  Ontology Identifier  \n",
       "0          GO:0046935  \n",
       "1          GO:0040024  \n",
       "2          GO:0008340  \n",
       "3          GO:0008286  \n",
       "4          GO:0016301  \n",
       "5          GO:0005942  \n",
       "6          GO:0046854  \n",
       "7          GO:0016310  \n",
       "8          GO:0019901  \n",
       "9          GO:0043551  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query=service.new_query(\"Gene\")\n",
    "query.select(\"primaryIdentifier\",\"symbol\", \"ontologyAnnotations.ontologyTerm.name\", \n",
    "             \"ontologyAnnotations.ontologyTerm.identifier\")\n",
    "query.add_constraint(\"homologues.type\",\"=\",\"orthologue\")\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "genes_data = pd.DataFrame(columns = [\"primaryIdentifier\", \"symbol\", \"Ontology Name\", \"Ontology Identifier\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'primaryIdentifier':row[0], 'symbol':row[1], 'Ontology Name':row[2], 'Ontology Identifier':row[3]}\n",
    "    genes_data = genes_data.append(info, ignore_index = True)\n",
    "genes_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is possible to perform INNER and OUTER joins on the queries to get columns from different sets of data easily!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>primaryIdentifier</th>\n",
       "      <th>symbol</th>\n",
       "      <th>Ontology Name</th>\n",
       "      <th>Ontology Identifier</th>\n",
       "      <th>Homologue Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>1-phosphatidylinositol-3-kinase regulator acti...</td>\n",
       "      <td>GO:0046935</td>\n",
       "      <td>orthologue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>1-phosphatidylinositol-3-kinase regulator acti...</td>\n",
       "      <td>GO:0046935</td>\n",
       "      <td>orthologue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>1-phosphatidylinositol-3-kinase regulator acti...</td>\n",
       "      <td>GO:0046935</td>\n",
       "      <td>orthologue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>1-phosphatidylinositol-3-kinase regulator acti...</td>\n",
       "      <td>GO:0046935</td>\n",
       "      <td>orthologue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>1-phosphatidylinositol-3-kinase regulator acti...</td>\n",
       "      <td>GO:0046935</td>\n",
       "      <td>orthologue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>1-phosphatidylinositol-3-kinase regulator acti...</td>\n",
       "      <td>GO:0046935</td>\n",
       "      <td>orthologue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>1-phosphatidylinositol-3-kinase regulator acti...</td>\n",
       "      <td>GO:0046935</td>\n",
       "      <td>orthologue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>1-phosphatidylinositol-3-kinase regulator acti...</td>\n",
       "      <td>GO:0046935</td>\n",
       "      <td>orthologue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>1-phosphatidylinositol-3-kinase regulator acti...</td>\n",
       "      <td>GO:0046935</td>\n",
       "      <td>orthologue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WBGene00000001</td>\n",
       "      <td>aap-1</td>\n",
       "      <td>1-phosphatidylinositol-3-kinase regulator acti...</td>\n",
       "      <td>GO:0046935</td>\n",
       "      <td>orthologue</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  primaryIdentifier symbol                                      Ontology Name  \\\n",
       "0    WBGene00000001  aap-1  1-phosphatidylinositol-3-kinase regulator acti...   \n",
       "1    WBGene00000001  aap-1  1-phosphatidylinositol-3-kinase regulator acti...   \n",
       "2    WBGene00000001  aap-1  1-phosphatidylinositol-3-kinase regulator acti...   \n",
       "3    WBGene00000001  aap-1  1-phosphatidylinositol-3-kinase regulator acti...   \n",
       "4    WBGene00000001  aap-1  1-phosphatidylinositol-3-kinase regulator acti...   \n",
       "5    WBGene00000001  aap-1  1-phosphatidylinositol-3-kinase regulator acti...   \n",
       "6    WBGene00000001  aap-1  1-phosphatidylinositol-3-kinase regulator acti...   \n",
       "7    WBGene00000001  aap-1  1-phosphatidylinositol-3-kinase regulator acti...   \n",
       "8    WBGene00000001  aap-1  1-phosphatidylinositol-3-kinase regulator acti...   \n",
       "9    WBGene00000001  aap-1  1-phosphatidylinositol-3-kinase regulator acti...   \n",
       "\n",
       "  Ontology Identifier Homologue Type  \n",
       "0          GO:0046935     orthologue  \n",
       "1          GO:0046935     orthologue  \n",
       "2          GO:0046935     orthologue  \n",
       "3          GO:0046935     orthologue  \n",
       "4          GO:0046935     orthologue  \n",
       "5          GO:0046935     orthologue  \n",
       "6          GO:0046935     orthologue  \n",
       "7          GO:0046935     orthologue  \n",
       "8          GO:0046935     orthologue  \n",
       "9          GO:0046935     orthologue  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query.add_view('homologues.type')\n",
    "query.add_join(\"homologues\", \"INNER\")\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "genes_data = pd.DataFrame(columns = [\"primaryIdentifier\", \"symbol\", \"Ontology Name\", \"Ontology Identifier\",\n",
    "                                    \"Homologue Type\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'primaryIdentifier':row[0], 'symbol':row[1], 'Ontology Name':row[2], 'Ontology Identifier':row[3], \n",
    "            'Homologue Type':row[4]}\n",
    "    genes_data = genes_data.append(info, ignore_index = True)\n",
    "genes_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Combinations of constraints and set logic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's query the organism name and gene symbol for all genes in the WormMine database. We then constrain our results based on the logic OR of two specified constraints as in the cell below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene.organism.name</th>\n",
       "      <th>Gene.symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>sgn-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>ugt-59</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene.organism.name Gene.symbol\n",
       "0  Caenorhabditis elegans       sgn-1\n",
       "1  Caenorhabditis elegans      ugt-59"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = service.new_query()\n",
    "query.add_view(\"Gene.organism.name\",\"Gene.symbol\")\n",
    "\n",
    "gene_is_ugt = query.add_constraint(\"Gene.symbol\", \"=\", \"ugt-59\") #Add first binary constraint\n",
    "gene_is_sgn = query.add_constraint(\"Gene.symbol\", \"=\", \"sgn-1\") #Add second binary constraint\n",
    "query.set_logic(gene_is_ugt | gene_is_sgn) #Logic OR on the 2 constraints\n",
    "\n",
    "#Insert first 10 rows of the query results into a dataframe and display the output!\n",
    "genes_data = pd.DataFrame(columns = [\"Gene.organism.name\", \"Gene.symbol\"])\n",
    "for row in query.rows(start=0,size=10):\n",
    "    info = {'Gene.organism.name':row[0], 'Gene.symbol':row[1]}\n",
    "    genes_data = genes_data.append(info, ignore_index = True)\n",
    "genes_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing query results to a file for later use"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since we use dataframes for storing the data, we can easily choose any rows or columns we want to retain based on simple constraints using the various functionalities present in the pandas library!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_data.to_csv('results.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting a readable XML serialisation of a query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<query name=\"\" model=\"genomic\" view=\"Gene.organism.name Gene.symbol\" sortOrder=\"Gene.organism.name asc\" longDescription=\"\" constraintLogic=\"A or B\"><constraint path=\"Gene.symbol\" op=\"=\" code=\"A\" value=\"ugt-59\"/><constraint path=\"Gene.symbol\" op=\"=\" code=\"B\" value=\"sgn-1\"/></query>'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query.to_xml()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clearing the output of a query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "query.clear_view()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is the end of the tutorial for querying and extracting WormBase data using WormMine through intermine. This tutorial is influenced by the intermine tutorial notebooks from - https://github.com/intermine/intermine-ws-python-docs\n",
    "\n",
    "In the next tutorial, we will use access the WormBase ParaSite data through their RESTful API."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
