{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Bioinformatics with Jupyter Notebooks for WormBase:**\n",
    "## **Data Retrieval 4 - Getting essential gene information**\n",
    "Welcome to the fourth jupyter notebook in the WormBase tutorial series. Over this series of tutorials, we will write code in Python that allows us to retrieve and perform simple analyses with data available on the WormBase sites.\n",
    "\n",
    "This tutorial will deal with getting the essential gene information for WormBase genes, replicating the results from the SimpleMine utility. Given a list of WormBase Gene IDs, we can extract all required information associated with the gene using the WormBase RESTful API. Let's get started!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We start by installing and loading the libraries that are required for this tutorial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests, sys\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None) #for ensuring full view of the dataframe generated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We initialise the columns for the dataframe. All columns have a small description next to them. You can comment out any of the columns if they are not required for your study."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "EssentialGeneInfo = pd.DataFrame(columns = ['WormBase Gene ID', #Unique Gene identifiers used by WormBase\n",
    "                                            'Public Name', #Official gene names specified by WormBase\n",
    "                                            'Species', #Each gene can only be associated with one species\n",
    "                                            'Sequence Name', #Sequence name of the gene\n",
    "                                            'Other Name', #All names that have been used by the gene in publications\n",
    "                                            'Transcript', #Transcript names of the gene\n",
    "                                            'Operon', #A set of genes transcribed under the control of an operator gene\n",
    "                                            'Protein Domain', #Protein Domains associated with the gene\n",
    "                                            'UniProt', #Official Protein Identifiers used by the UniProt database\n",
    "                                            'Reference UniProt ID', #Unique UniProt ID for each gene to act as a \"reference\" UniProt ID\n",
    "                                            'TreeFam', #Official gene identifiers used by the TreeFam database\n",
    "                                            'RefSeq_mRNA', #Sequence IDs used by the RefSeq database\n",
    "                                            'RefSeq_protein', #Protein IDs used by the RefSeq databaseDisplay Chromosome and chromosomal position of the gene\n",
    "                                            'Genetic Map Position', #Display Chromosome and chromosomal position of the gene\n",
    "                                            'RNAi Phenotype Observed', #Display the RNAi phenotype ontology names\n",
    "                                            'Allele Phenotype Observed', #Display the Allele phenotype ontology names\n",
    "                                            'Coding_exon Non_silent Allele', #Display a list of alleles that fall in any coding exon\n",
    "                                            'Interacting Gene', #Display experimentally confirmed gene interactions\n",
    "                                            'Expr_pattern Tissue', #Anatomical expression based on GFP, immunoprecipitation, In_situ\n",
    "                                            'Genomic Study Tissue', #Tissue enrichment based on the microarray, RNA-Seq, and proteomics studies\n",
    "                                            'Expr_pattern LifeStage', #Developmental expression based on GFP, immunoprecipitation, In_situ\n",
    "                                            'Genomic Study LifeStage', #Developmental expression based on the microarray, RNA-Seq, and proteomics studies\n",
    "                                            'Disease Info', #Display the disease names associated with the gene\n",
    "                                            'Human Ortholog', #Display the human orthologs of the gene\n",
    "                                            'Gene Ontology Association', #Display the names of gene ontology terms that were annotated to the gene\n",
    "                                            'Automated Description', #Up-to-date gene description \n",
    "                                            'Reference' #Primary research articles that studied the gene\n",
    "                                           ])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next step is to initialize the list of Gene IDs. This can be done either by manually typing in a list or by uploading a csv file that has one single column with the required gene IDs.\n",
    "Uncomment the required line in the below cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "GeneID = ['WBGene00001648', 'WBGene00012578', 'WBGene00021277']\n",
    "#GeneID = pd.read_csv('GeneID_list.csv', header=None)[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We finally access the WormBase RESTful API to get the necessary information for the genes. We use the gene widgets and gene fields operations to access the data based on the gene ID. (http://rest.wormbase.org/index.html)\n",
    "\n",
    "Remember to comment out any fields that you may have commented in the previous cell while initialising the empty dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in GeneID:\n",
    "    WormBaseGeneID = gene\n",
    "    PublicName = requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/name').json()['name']['data']['label']\n",
    "    Species = requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/taxonomy').json()['taxonomy']['data']['genus'] + ' ' + requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/taxonomy').json()['taxonomy']['data']['species']\n",
    "    SequenceName = requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/sequence_name').json()['sequence_name']['data']\n",
    "    OtherName = ', '. join(requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/other_names').json()['other_names']['data'])\n",
    "    Transcript = requests.get('http://rest.wormbase.org/rest/widget/gene/' + gene + '/sequences').json()['fields']['gene_models']['data']['table'][0]['model'][0]['id']\n",
    "    Operon = 'N.A.' if requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/operon').json()['operon']['data'] is None else requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/operon').json()['operon']['data']['label']\n",
    "    ProteinDomain = [val['id'] for key, val in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/protein_domains').json()['protein_domains']['data'].items() if 'id' in val]\n",
    "    Uniprot = requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/xrefs').json()['xrefs']['data']['TrEMBL']['UniProtAcc']['ids'] if 'TrEMBL' in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/xrefs').json()['xrefs']['data'] else requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/xrefs').json()['xrefs']['data']['SwissProt']['UniProtAcc']['ids'] if 'SwissProt' in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/xrefs').json()['xrefs']['data'] else 'N.A.'\n",
    "    RefUniprotID = requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/xrefs').json()['xrefs']['data']['UniProt_GCRP']['UniProtAcc']['ids']\n",
    "    TreeFam = 'N.A.' if requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/treefam').json()['treefam']['data'] is None else requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/treefam').json()['treefam']['data']\n",
    "    RefSeqmRNA = requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/xrefs').json()['xrefs']['data']['RefSeq']['mRNA']['ids']\n",
    "    RefSeqProtein = requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/xrefs').json()['xrefs']['data']['RefSeq']['protein']['ids']\n",
    "    GeneticMapPosition = requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/genetic_position').json()['genetic_position']['data'][0]['chromosome'] + ':' + str(requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/genetic_position').json()['genetic_position']['data'][0]['position'])\n",
    "    RNAiPhen = sorted(list(set([i['phenotype']['label'] for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/phenotype').json()['phenotype']['data'] if 'RNAi' in i['evidence']])))\n",
    "    AllelePhen = sorted(list(set([i['phenotype']['label'] for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/phenotype').json()['phenotype']['data'] if 'Allele' in i['evidence']])))\n",
    "    NonSilent = sorted(list(set([i['variation']['label'] + '|' + i['effects'][0] for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/alleles').json()['alleles']['data'] if 'effects' in i])))\n",
    "    InteractingGene = sorted(list(set([i['affected']['label'] for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/interaction_details').json()['interaction_details']['data']['edges_all']])))\n",
    "    ExprPatternTissue = sorted([i['ontology_term']['label'] for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/expressed_in').json()['expressed_in']['data'] if i['details'][0]['text']['class'] == 'expr_pattern'])\n",
    "    GenomicStudyTissue = sorted([i['ontology_term']['label'] for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/expressed_in').json()['expressed_in']['data'] if i['details'][0]['text']['class'] == 'expression_cluster'])\n",
    "    ExprPatternLifeStage = 'N.A.' if requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/expressed_during').json()['expressed_during']['data'] is None else sorted([i['ontology_term']['label'] for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/expressed_during').json()['expressed_during']['data']])\n",
    "    GenomicStudyLifeStage = 'N.A.' if requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/fpkm_expression_summary_ls').json()['fpkm_expression_summary_ls']['data']['table']['fpkm']['data'] is None else sorted(list(set([i['life_stage']['label'] for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/fpkm_expression_summary_ls').json()['fpkm_expression_summary_ls']['data']['table']['fpkm']['data']])))\n",
    "    Disease = 'N.A.' if requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/human_diseases').json()['human_diseases']['data'] is None else sorted([i['label'] for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/human_diseases').json()['human_diseases']['data']['potential_model']])\n",
    "    HumanOrtholog = 'N.A.' if requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/other_orthologs').json()['other_orthologs']['data'] is None else sorted([i['ortholog']['id'] + '|' + '; '.join([j['id'] for j in i['method']]) for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/other_orthologs').json()['other_orthologs']['data'] if (i['ortholog']['taxonomy'] == 'h_sapiens')])\n",
    "    GeneOntology = 'N.A.' if requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/gene_ontology').json()['gene_ontology']['data'] is None else sorted(list(set([i['term_description']['label'] for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/gene_ontology').json()['gene_ontology']['data']['Biological_process']] + [i['term_description']['label'] for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/gene_ontology').json()['gene_ontology']['data']['Molecular_function']] + [i['term_description']['label'] for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/gene_ontology').json()['gene_ontology']['data']['Cellular_component']] + [i['with'][0]['label'] for i in requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/gene_ontology').json()['gene_ontology']['data']['Cellular_component'] if i['with'] is not None and i['with'][0]['label'][:7] != 'Panther'])))\n",
    "    AutomatedDesc = requests.get('http://rest.wormbase.org/rest/field/gene/' + gene + '/concise_description').json()['concise_description']['data']['text']\n",
    "    Reference = 'N.A.' if requests.get('http://rest.wormbase.org/rest/widget/gene/' + gene + '/references').json()['fields']['references']['data'] is None else [i['name']['id'] for i in requests.get('http://rest.wormbase.org/rest/widget/gene/' + gene + '/references').json()['fields']['references']['data']['results']]\n",
    "    EssentialGeneInfo = EssentialGeneInfo.append({'WormBase Gene ID': WormBaseGeneID, \n",
    "                                            'Public Name': PublicName,\n",
    "                                            'Species': Species,\n",
    "                                            'Sequence Name': SequenceName,\n",
    "                                            'Other Name': OtherName,\n",
    "                                            'Transcript': Transcript,\n",
    "                                            'Operon': Operon,\n",
    "                                            'Protein Domain': ProteinDomain, \n",
    "                                            'UniProt': Uniprot,\n",
    "                                            'Reference UniProt ID': RefUniprotID,\n",
    "                                            'TreeFam': TreeFam,\n",
    "                                            'RefSeq_mRNA': RefSeqmRNA,\n",
    "                                            'RefSeq_protein': RefSeqProtein,\n",
    "                                            'Genetic Map Position': GeneticMapPosition,\n",
    "                                            'RNAi Phenotype Observed': RNAiPhen,\n",
    "                                            'Allele Phenotype Observed': AllelePhen,\n",
    "                                            'Coding_exon Non_silent Allele': NonSilent,\n",
    "                                            'Interacting Gene': InteractingGene,\n",
    "                                            'Expr_pattern Tissue': ExprPatternTissue,\n",
    "                                            'Genomic Study Tissue': GenomicStudyTissue,\n",
    "                                            'Expr_pattern LifeStage': ExprPatternLifeStage,\n",
    "                                            'Genomic Study LifeStage': GenomicStudyLifeStage,\n",
    "                                            'Disease Info': Disease,\n",
    "                                            'Human Ortholog': HumanOrtholog,\n",
    "                                            'Gene Ontology Association': GeneOntology,\n",
    "                                            'Automated Description': AutomatedDesc,\n",
    "                                            'Reference': Reference}, \n",
    "                                            ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>WormBase Gene ID</th>\n",
       "      <th>Public Name</th>\n",
       "      <th>Species</th>\n",
       "      <th>Sequence Name</th>\n",
       "      <th>Other Name</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Operon</th>\n",
       "      <th>Protein Domain</th>\n",
       "      <th>UniProt</th>\n",
       "      <th>Reference UniProt ID</th>\n",
       "      <th>TreeFam</th>\n",
       "      <th>RefSeq_mRNA</th>\n",
       "      <th>RefSeq_protein</th>\n",
       "      <th>Genetic Map Position</th>\n",
       "      <th>RNAi Phenotype Observed</th>\n",
       "      <th>Allele Phenotype Observed</th>\n",
       "      <th>Coding_exon Non_silent Allele</th>\n",
       "      <th>Interacting Gene</th>\n",
       "      <th>Expr_pattern Tissue</th>\n",
       "      <th>Genomic Study Tissue</th>\n",
       "      <th>Expr_pattern LifeStage</th>\n",
       "      <th>Genomic Study LifeStage</th>\n",
       "      <th>Disease Info</th>\n",
       "      <th>Human Ortholog</th>\n",
       "      <th>Gene Ontology Association</th>\n",
       "      <th>Automated Description</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WBGene00001648</td>\n",
       "      <td>goa-1</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>C26C6.2</td>\n",
       "      <td>unc-109, CELE_C26C6.2</td>\n",
       "      <td>C26C6.2.1</td>\n",
       "      <td>N.A.</td>\n",
       "      <td>[INTERPRO:IPR001408, INTERPRO:IPR027417, INTER...</td>\n",
       "      <td>[P51875]</td>\n",
       "      <td>[P51875]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[NM_059707.6]</td>\n",
       "      <td>[NP_492108.1]</td>\n",
       "      <td>I:2.098785</td>\n",
       "      <td>[P0 spindle position defective early emb, anti...</td>\n",
       "      <td>[acetylcholine synaptic transmission variant, ...</td>\n",
       "      <td>[n1134|Missense, n499|Missense, sa734|Nonsense...</td>\n",
       "      <td>[F10D7.5, F38H4.3, F47F2.1, F53A10.2, W02B3.8,...</td>\n",
       "      <td>[ALML, ALMR, AVM, BDUL, BDUR, CANL, CANR, HSNL...</td>\n",
       "      <td>[AFD, ASER, AVK, DA neuron, NSM, OLL, PVD, RIS...</td>\n",
       "      <td>[adult Ce, embryo Ce, larva Ce, postembryonic Ce]</td>\n",
       "      <td>[1-cell embryo Ce, 1-day post-L4 adult hermaph...</td>\n",
       "      <td>[developmental and epileptic encephalopathy 17]</td>\n",
       "      <td>[HGNC:4389|EnsEMBL-Compara; OrthoFinder; Round...</td>\n",
       "      <td>[G protein-coupled acetylcholine receptor acti...</td>\n",
       "      <td>Enables several functions, including G protein...</td>\n",
       "      <td>[WBPaper00014801, WBPaper00034090, WBPaper0003...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WBGene00012578</td>\n",
       "      <td>ccct-1</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Y37H9A.3</td>\n",
       "      <td>CELE_Y37H9A.3</td>\n",
       "      <td>Y37H9A.3.1</td>\n",
       "      <td>CEOP1933</td>\n",
       "      <td>[INTERPRO:IPR035892, INTERPRO:IPR039725, INTER...</td>\n",
       "      <td>[Q9U2M8]</td>\n",
       "      <td>[Q9U2M8]</td>\n",
       "      <td>[TF314229]</td>\n",
       "      <td>[NM_001373656.2]</td>\n",
       "      <td>[NP_001359623.1]</td>\n",
       "      <td>I:21.6609</td>\n",
       "      <td>[sterile]</td>\n",
       "      <td>[lethal, sterile]</td>\n",
       "      <td>[tm2372|Splice site]</td>\n",
       "      <td>[C30F12.4, F07F6.8, F59D12.2, R119.2, ccct-1, ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[NSM, germ line, intestine]</td>\n",
       "      <td>N.A.</td>\n",
       "      <td>[1-cell embryo Ce, 1-day post-L4 adult hermaph...</td>\n",
       "      <td>[autosomal recessive non-syndromic intellectua...</td>\n",
       "      <td>[HGNC:29386|EnsEMBL-Compara; OrthoFinder; Inpa...</td>\n",
       "      <td>[DNA-binding transcription factor activity, RN...</td>\n",
       "      <td>Is predicted to enable DNA-binding transcripti...</td>\n",
       "      <td>[WBPaper00038491, WBPaper00055090]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WBGene00021277</td>\n",
       "      <td>ddx-10</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Y23H5B.6</td>\n",
       "      <td>CELE_Y23H5B.6, Y74C10A_153.a, Y23H5B.c</td>\n",
       "      <td>Y23H5B.6a.1</td>\n",
       "      <td>CEOP1048</td>\n",
       "      <td>[INTERPRO:IPR027417, INTERPRO:IPR025313, INTER...</td>\n",
       "      <td>[Q9N478, A0A061ADT4, A0A061ACL9]</td>\n",
       "      <td>[Q9N478]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[NM_001306361.3, NM_058588.6, NM_001306360.3]</td>\n",
       "      <td>[NP_001293290.1, NP_490989.1, NP_001293289.1]</td>\n",
       "      <td>I:-8.47867</td>\n",
       "      <td>[embryonic lethal, larval arrest, late larval ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[B0280.9, B0511.6, F57B10.8, F58B3.4, K01G5.5,...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[AVE, body wall muscle cell, germ line, intest...</td>\n",
       "      <td>N.A.</td>\n",
       "      <td>[1-cell embryo Ce, 1-day post-L4 adult hermaph...</td>\n",
       "      <td>N.A.</td>\n",
       "      <td>[HGNC:2735|EnsEMBL-Compara; OrthoFinder; Round...</td>\n",
       "      <td>[ATP binding, RNA binding, RNA helicase activi...</td>\n",
       "      <td>Is predicted to enable ATP binding activity; R...</td>\n",
       "      <td>[WBPaper00038491, WBPaper00055090]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  WormBase Gene ID Public Name                 Species Sequence Name  \\\n",
       "0   WBGene00001648       goa-1  Caenorhabditis elegans       C26C6.2   \n",
       "1   WBGene00012578      ccct-1  Caenorhabditis elegans      Y37H9A.3   \n",
       "2   WBGene00021277      ddx-10  Caenorhabditis elegans      Y23H5B.6   \n",
       "\n",
       "                               Other Name   Transcript    Operon  \\\n",
       "0                   unc-109, CELE_C26C6.2    C26C6.2.1      N.A.   \n",
       "1                           CELE_Y37H9A.3   Y37H9A.3.1  CEOP1933   \n",
       "2  CELE_Y23H5B.6, Y74C10A_153.a, Y23H5B.c  Y23H5B.6a.1  CEOP1048   \n",
       "\n",
       "                                      Protein Domain  \\\n",
       "0  [INTERPRO:IPR001408, INTERPRO:IPR027417, INTER...   \n",
       "1  [INTERPRO:IPR035892, INTERPRO:IPR039725, INTER...   \n",
       "2  [INTERPRO:IPR027417, INTERPRO:IPR025313, INTER...   \n",
       "\n",
       "                            UniProt Reference UniProt ID     TreeFam  \\\n",
       "0                          [P51875]             [P51875]          []   \n",
       "1                          [Q9U2M8]             [Q9U2M8]  [TF314229]   \n",
       "2  [Q9N478, A0A061ADT4, A0A061ACL9]             [Q9N478]          []   \n",
       "\n",
       "                                     RefSeq_mRNA  \\\n",
       "0                                  [NM_059707.6]   \n",
       "1                               [NM_001373656.2]   \n",
       "2  [NM_001306361.3, NM_058588.6, NM_001306360.3]   \n",
       "\n",
       "                                  RefSeq_protein Genetic Map Position  \\\n",
       "0                                  [NP_492108.1]           I:2.098785   \n",
       "1                               [NP_001359623.1]            I:21.6609   \n",
       "2  [NP_001293290.1, NP_490989.1, NP_001293289.1]           I:-8.47867   \n",
       "\n",
       "                             RNAi Phenotype Observed  \\\n",
       "0  [P0 spindle position defective early emb, anti...   \n",
       "1                                          [sterile]   \n",
       "2  [embryonic lethal, larval arrest, late larval ...   \n",
       "\n",
       "                           Allele Phenotype Observed  \\\n",
       "0  [acetylcholine synaptic transmission variant, ...   \n",
       "1                                  [lethal, sterile]   \n",
       "2                                                 []   \n",
       "\n",
       "                       Coding_exon Non_silent Allele  \\\n",
       "0  [n1134|Missense, n499|Missense, sa734|Nonsense...   \n",
       "1                               [tm2372|Splice site]   \n",
       "2                                                 []   \n",
       "\n",
       "                                    Interacting Gene  \\\n",
       "0  [F10D7.5, F38H4.3, F47F2.1, F53A10.2, W02B3.8,...   \n",
       "1  [C30F12.4, F07F6.8, F59D12.2, R119.2, ccct-1, ...   \n",
       "2  [B0280.9, B0511.6, F57B10.8, F58B3.4, K01G5.5,...   \n",
       "\n",
       "                                 Expr_pattern Tissue  \\\n",
       "0  [ALML, ALMR, AVM, BDUL, BDUR, CANL, CANR, HSNL...   \n",
       "1                                                 []   \n",
       "2                                                 []   \n",
       "\n",
       "                                Genomic Study Tissue  \\\n",
       "0  [AFD, ASER, AVK, DA neuron, NSM, OLL, PVD, RIS...   \n",
       "1                        [NSM, germ line, intestine]   \n",
       "2  [AVE, body wall muscle cell, germ line, intest...   \n",
       "\n",
       "                              Expr_pattern LifeStage  \\\n",
       "0  [adult Ce, embryo Ce, larva Ce, postembryonic Ce]   \n",
       "1                                               N.A.   \n",
       "2                                               N.A.   \n",
       "\n",
       "                             Genomic Study LifeStage  \\\n",
       "0  [1-cell embryo Ce, 1-day post-L4 adult hermaph...   \n",
       "1  [1-cell embryo Ce, 1-day post-L4 adult hermaph...   \n",
       "2  [1-cell embryo Ce, 1-day post-L4 adult hermaph...   \n",
       "\n",
       "                                        Disease Info  \\\n",
       "0    [developmental and epileptic encephalopathy 17]   \n",
       "1  [autosomal recessive non-syndromic intellectua...   \n",
       "2                                               N.A.   \n",
       "\n",
       "                                      Human Ortholog  \\\n",
       "0  [HGNC:4389|EnsEMBL-Compara; OrthoFinder; Round...   \n",
       "1  [HGNC:29386|EnsEMBL-Compara; OrthoFinder; Inpa...   \n",
       "2  [HGNC:2735|EnsEMBL-Compara; OrthoFinder; Round...   \n",
       "\n",
       "                           Gene Ontology Association  \\\n",
       "0  [G protein-coupled acetylcholine receptor acti...   \n",
       "1  [DNA-binding transcription factor activity, RN...   \n",
       "2  [ATP binding, RNA binding, RNA helicase activi...   \n",
       "\n",
       "                               Automated Description  \\\n",
       "0  Enables several functions, including G protein...   \n",
       "1  Is predicted to enable DNA-binding transcripti...   \n",
       "2  Is predicted to enable ATP binding activity; R...   \n",
       "\n",
       "                                           Reference  \n",
       "0  [WBPaper00014801, WBPaper00034090, WBPaper0003...  \n",
       "1                 [WBPaper00038491, WBPaper00055090]  \n",
       "2                 [WBPaper00038491, WBPaper00055090]  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "EssentialGeneInfo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The dataframe can be written into a csv file to save the essential gene information and for any analyses that you want to perform later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "EssentialGeneInfo.to_csv('EssentialGeneInformation.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is the end of the tutorial on replicating SimpleMine results using the WormBase RESTful API to get the essential gene information. The data is up-to date and is very quick to extract, and is easier to handle than the results from SimpleMine.\n",
    "\n",
    "This tutorial is also the end of the Data Retrieval series. In the next tutorial, we will implement and test some simple utilities that can help us work with the data we have retrieved until now.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
